### PR TITLE
Add history selection to examples

### PR DESCRIPTION
**Pull request by kmadsen**
## Description

Making it so we can select many history files for debugging.

First we need a public repository to share history files https://github.com/mapbox/mapbox-navigation-history/pull/1

- [ ] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

## Screenshots or Gifs

| selecting different history files |
| -- |
| output |
## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt [hunk 1]
```diff
@@ -106,6 +119,23 @@ class ReplayHistoryActivity : AppCompatActivity() {
...
+        HistoryFilesActivity.selectedHistory?.let {
+            val replayHistoryMapper = ReplayHistoryMapper(Gson(), ReplayCustomEventMapper(), MapboxLogger)
+            val replayEvents = replayHistoryMapper.mapToReplayEvents(it)
+            navigationContext?.mapboxReplayer?.stop()
+            navigationContext?.mapboxReplayer?.clearEvents()
+            navigationContext?.mapboxReplayer?.pushEvents(replayEvents)
+            navigationContext?.mapboxReplayer?.seekTo(0.0)
+
+            // We have a new set of data, so reset the navigator
+            MapboxNativeNavigatorImpl.create(HandheldProfile(), MapboxLogger)
```
Comment (kmadsen) [inline examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt:134]: There is offline discussion on how we should handle this. The only way we know to reset navigation after selecting history files, is to use an internal method. This call can also be triggered with a public function using MapboxNavigation.onDestroy

**Comment by Guardiola31337** [comment]
Testing `ReplayHistoryActivity` run into the same 💥 in https://github.com/mapbox/mapbox-navigation-android/pull/3143#issue-432114069 but using `13.0.2` 🤔 

```
06-11 09:57:39.763 16559-16568/? E/AndroidRuntime: FATAL EXCEPTION: FinalizerWatchdogDaemon
    Process: com.mapbox.navigation.examples, PID: 16559
    java.util.concurrent.TimeoutException: com.mapbox.navigator.Navigator.finalize() timed out after 10 seconds
        at com.mapbox.navigator.Navigator.finalize(Native Method)
        at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:202)
        at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:185)
        at java.lang.Thread.run(Thread.java:818)
```

Hopefully fixed with the upcoming NN version 13.x Refs. https://github.com/mapbox/mapbox-navigation-android/pull/3143 https://github.com/mapbox/mapbox-navigation-android/pull/3163

cc @mskurydin

**Comment by Guardiola31337** [comment]
@kmadsen Took this for a spin and noticed that `ReplayHistoryActivity` starts in _Free Drive_ and stays in that mode even after selecting a history file 👀 

replay_history_activity_free_drive

Is that expected? I believe it relates to `unsupported event setRoute` logs 👇 

```
06-12 11:55:43.674 20392-20392/com.mapbox.navigation.examples E/MapboxLogger: Replay unsupported event setRoute
```

I'm seeing in the Logcat.

**Comment by kmadsen** [comment]
> @ kmadsen Took this for a spin and noticed that ReplayHistoryActivity starts in Free Drive and stays in that mode even after selecting a history file 👀

The definition for ReplayHistoryActivity is to replicate the state of the history file. History starts when we start a session (free drive or active guidance).

I think we should make it start Free Drive after you select MapboxReplayer.play(). Because we're not mapping all events yet, ReplayHistoryActivity is still transitioning to full completion. I'm not sure if we're able to fix the Free Drive vs Active Guidance, until we add the `setRoute` event.

But adding setRoute can come after this - I added the change but found it will require a few more changes and tests.

**Comment by kmadsen** [comment]
> Did we rebase from master? Refs. #3178

Yeah was rebased, shouldn't be crashing. I'm not getting that stack trace, are you still able to repro? MaterialCard's have different depending on the android version? 🤔  😆

**Comment by kmadsen** [comment]
> Also, one thing that'd be great to add in the ReplayHistoryActivity is moving the camera / map to the first location from the history file after selecting a new one.

This was also fixed, but maybe still reproducible? Am relying on the navigator to update the camera and location, which is why there is this reset. 
```
// We have a new set of data, so reset the navigator
MapboxNativeNavigatorImpl.create(HandheldProfile(), MapboxLogger)
```

May be another reason to build the "location repeater" for replay. It also provides a resolution here https://github.com/mapbox/mapbox-navigation-native/issues/1249

**Comment by Guardiola31337** [comment]
@kmadsen 
> Yeah was rebased, shouldn't be crashing. I'm not getting that stack trace, are you still able to repro? MaterialCard's have different depending on the android version? 🤔 😆

Not able to reproduce anymore 🎉 

> This was also fixed, but maybe still reproducible?

To clarify here, what I meant was to move the camera after loading the history file - after clicking on `START NAVIGATION` camera is moved to the proper location but if not puck is displayed in 🇯🇵 (history file that is loaded by default). Does that make sense?

File: examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    @SerializedName("description")
+    val description: String,
+    @SerializedName("path")
+    val path: String
+)
+
+class HistoryFilesClient {
+
+    private val retrofit = Retrofit.Builder()
+        .baseUrl("https://mapbox.github.io/driver/replay-history/")
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt:28]: Are we 🆗 with exposing this? cc @zugaldia

**Comment by Guardiola31337** [comment]
> Also, one thing that'd be great to add in the `ReplayHistoryActivity` is moving the camera / map to the first location from the history file after selecting a new one.

Noting that this is fixed now 🚀

File: examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt (hunk 2 above)
Comment (Guardiola31337) [reply examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt:28]: Noting that this Github page is publicly accessible, not sure about the security implications though.

Comment (kmadsen) [reply examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt:28]: yeah, in agreement. we'll make a solution that is deliberate https://github.com/mapbox/navigation-sdks/issues/412

Comment (kmadsen) [reply examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt]: OK cool, the new pages are up

https://github.com/mapbox/mapbox-navigation-history
Available history files: https://mapbox.github.io/mapbox-navigation-history/index.json
Selected files: https://mapbox.github.io/mapbox-navigation-history/navigation-history/2020-02-04-germany-highway.json

Deprecated driver reference has been taken down
https://mapbox.github.io/driver/replay-history/index.json

Comment (zugaldia) [reply examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt]: cc: @1ec5 I assume they're generic enough that iOS will also be able to take advantage of this set up.

File: examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt [hunk 3]
```diff
@@ -75,9 +88,8 @@ class ReplayHistoryActivity : AppCompatActivity() {
             // Load map and style on Main dispatchers
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt:88]: NIT unnecessary comment

File: examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt [hunk 4]
```diff
@@ -75,9 +88,8 @@ class ReplayHistoryActivity : AppCompatActivity() {
...
             val deferredMapboxWithStyle = async { loadMapWithStyle() }
 
-            // Load and replay history on IO dispatchers
-            val deferredEvents = async(Dispatchers.IO) { loadReplayHistory() }
-            val replayEvents = deferredEvents.await()
+            // Load and replay history
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt:91]: NIT unnecessary comment

File: examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesActivity.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    override fun onBackPressed() {
+        super.onBackPressed()
+
+        launchReplayHistory()
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        // clear the selected history
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesActivity.kt:64]: NIT unnecessary comment

File: examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    @SerializedName("description")
+    val description: String,
+    @SerializedName("path")
+    val path: String
+)
+
+class HistoryFilesClient {
+
+    private val retrofit = Retrofit.Builder()
+        .baseUrl("https://mapbox.github.io/mapbox-navigation-history/")
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesClient.kt:28]: Could we extract these magic numbers into constants?

File: examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesViewController.kt [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    private suspend fun requestHistoryDisk(context: Context): List<ReplayPath> = withContext(Dispatchers.IO) {
+        val historyFiles: List<String> = context.assets.list("")?.toList()
+            ?: Collections.emptyList()
+
+        historyFiles.filter { it.endsWith(".json") }
+            .map { fileName ->
+                ReplayPath(
+                    "Local history file",
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/history/HistoryFilesViewController.kt:80]: More magic numbers 😅

**Pull request merged**